File: Sources/Client/Public/Sources/Utils/Color.ts
import { memo } from "./Memoize";

// #00000030
export const DARKENED_BASE = 0.1875;

/**
 * Type alias that represents color code.
 */
export type ColorCode = `#${string}${string}${string}`;

/**
 * Darkens colour.
 * 
 * @param color - Color code
 * @param strength - Strenth
 */
export const darkened = memo((color: ColorCode, strength: number): ColorCode => {
    let r = parseInt(color.slice(1, 3), 16);
    let g = parseInt(color.slice(3, 5), 16);
    let b = parseInt(color.slice(5, 7), 16);

    r = Math.floor(r * (1 - strength));
    g = Math.floor(g * (1 - strength));
    b = Math.floor(b * (1 - strength));

    return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}` satisfies ColorCode;
});

/**
 * ligtens colour.
 * 
 * @param color - Color code
 * @param strength - Strenth
 */
export const lightened = memo((color: ColorCode, strength: number): ColorCode => {
    let r = parseInt(color.slice(1, 3), 16);
    let g = parseInt(color.slice(3, 5), 16);
    let b = parseInt(color.slice(5, 7), 16);
    
    r = Math.min(255, Math.floor(r * (1 + strength)));
    g = Math.min(255, Math.floor(g * (1 + strength)));
    b = Math.min(255, Math.floor(b * (1 + strength)));
    
    return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}` satisfies ColorCode;
});

File: Sources/zigclient/private/src/Dom/Canvas/Color.zig
const std = @import("std");
const Color = @This();

r: u8,
g: u8,
b: u8,

pub inline fn darkened(self: Color, comptime strength: f64) Color {
    const strength_c = comptime (1 - strength);

    return Color{
        .r = @floor(self.r * strength_c),
        .g = @floor(self.g * strength_c),
        .b = @floor(self.b * strength_c),
    };
}

pub inline fn lightened(self: Color, comptime strength: f64) Color {
    const strength_a = comptime (1 + strength);

    return Color{
        .r = @min(255, @floor(self.r * strength_a)),
        .g = @min(255, @floor(self.g * strength_a)),
        .b = @min(255, @floor(self.b * strength_a)),
    };
}

pub fn fromHex(comptime hex: *const [6:0]u8) Color {
    comptime {
        const r = std.fmt.parseInt(u8, hex[0..2], 16) catch unreachable;
        const g = std.fmt.parseInt(u8, hex[2..4], 16) catch unreachable;
        const b = std.fmt.parseInt(u8, hex[4..6], 16) catch unreachable;

        return Color{ .r = r, .g = g, .b = b };
    }
}
